server/service/src/item/item_variant/upsert/generate.rs

```diff
@@ -0,0 +1,127 @@
+use chrono::Utc;
+use repository::{
+    item_variant::{item_variant::ItemVariant, item_variant_row::ItemVariantRow},
+    ActivityLogType, RepositoryError,
+};
+
+use crate::{activity_log::activity_log_entry, service_provider::ServiceContext};
+
+use super::UpsertItemVariantWithPackaging;
+
+pub fn generate(
+    user_id: &str,
+    existing_variant: Option<ItemVariant>,
+    UpsertItemVariantWithPackaging {
+        id,
+        name,
+        item_id,
+        cold_storage_type_id,
+        manufacturer_id,
+        packaging_variants: _, // Mapped separately
+        doses_per_unit,
+        vvm_type,
+    }: UpsertItemVariantWithPackaging,
+) -> ItemVariantRow {
+    let (created_datetime, created_by) = match existing_variant {
+        Some(ref variant) => (
+            variant.item_variant_row.created_datetime,
+            variant.item_variant_row.created_by.clone(),
+        ),
+        None => (Utc::now().naive_utc(), Some(user_id.to_string())),
+    };
+
+    ItemVariantRow {
+        id,
+        name,
+        item_link_id: item_id,
+        cold_storage_type_id: cold_storage_type_id.map(|l| l.value).unwrap_or_default(),
+        manufacturer_link_id: manufacturer_id
+            .map(|manufacturer_id| manufacturer_id.value)
+            .unwrap_or_default(),
+        doses_per_unit,
+        vvm_type: vvm_type.map(|vvm_type| vvm_type.value).unwrap_or_default(),
+        created_datetime,
+        created_by,
+        deleted_datetime: None,
+    }
+}
+
+pub fn generate_logs(
+    ctx: &ServiceContext,
+    existing_variant: Option<ItemVariant>,
+    updated_variant: ItemVariant,
+) -> Result<(), RepositoryError> {
+    if let Some(existing_variant) = existing_variant {
+        let existing_item_variant = existing_variant.item_variant_row;
+        let updated_item_variant = updated_variant.item_variant_row;
+
+        if existing_item_variant.name != updated_item_variant.name {
+            activity_log_entry(
+                ctx,
+                ActivityLogType::ItemVariantUpdatedName,
+                Some(existing_item_variant.id.clone()),
+                Some(existing_item_variant.name.clone()),
+                Some(updated_item_variant.name.clone()),
+            )?;
+        }
+
+        if existing_item_variant.cold_storage_type_id != updated_item_variant.cold_storage_type_id {
+            if let (Some(existing_storage_type), Some(updated_storage_type)) = (
```

Comment:
Ohhh its bc the Some() Some()! Only creating a log when it existed previously, and you're changing it to something else, not removing.

We should also log when changing to or from None 👀

Reply:
Opps, don't know how I missed that >.< Thanks!